examples/biphenyl_torsion_sampling_hrex.py

```diff
@@ -0,0 +1,331 @@
+) -> List[BoundPotential]:
+    assert 0.0 <= lamb <= 1.0
+
+    if isinstance(top, BaseTopology):
+        return get_potentials_vacuum(top, ff_params, decouple_atom_idxs, lamb)
+    elif isinstance(top, HostGuestTopology):
+        return get_potentials_solvent(top, ff_params, decouple_atom_idxs, lamb)
+    else:
+        assert False
+
+
+def get_potentials_vacuum(
+    top: BaseTopology,
+    ff_params: ForcefieldParams,
+    decouple_atom_idxs: NDArray[np.int_],
+    lamb: float,
+) -> List[BoundPotential]:
+    hb_params, hb_pot = top.parameterize_harmonic_bond(ff_params.hb_params)
+    ha_params, ha_pot = top.parameterize_harmonic_angle(ff_params.ha_params)
+
+    ppt_params, ppt_pot = top.parameterize_proper_torsion(ff_params.pt_params)
+    ipt_params, ipt_pot = top.parameterize_improper_torsion(ff_params.it_params)
+    pt_params = np.concatenate([ppt_params, ipt_params])
+    pt_pot = PeriodicTorsion(np.concatenate([ppt_pot.idxs, ipt_pot.idxs]))
+    nb_params, nb_pot = top.parameterize_nonbonded_pairlist(
+        ff_params.q_params,
+        ff_params.q_params_intra,
+        ff_params.lj_params,
+        ff_params.lj_params_intra,
+    )
+
+    decouple_atom = np.isin(nb_pot.idxs, decouple_atom_idxs)
+    decouple_pair = np.logical_xor(decouple_atom[:, 0], decouple_atom[:, 1])
+    nb_params[decouple_pair, 3] = lamb
+
+    return [
+        hb_pot.bind(hb_params),
+        ha_pot.bind(ha_params),
+        pt_pot.bind(pt_params),
+        nb_pot.bind(nb_params),
+    ]
+
+
+def get_potentials_solvent(
+    top: HostGuestTopology,
+    ff_params: ForcefieldParams,
+    decouple_atom_idxs: NDArray[np.int_],
+    lamb: float,
+) -> List[BoundPotential]:
+    hb_params, hb_pot = top.parameterize_harmonic_bond(ff_params.hb_params)
+    ha_params, ha_pot = top.parameterize_harmonic_angle(ff_params.ha_params)
+
+    pt_params, pt_pot = top.parameterize_periodic_torsion(ff_params.pt_params, ff_params.it_params)
+    nb_params, nb_pot = top.parameterize_nonbonded(
+        ff_params.q_params,
+        ff_params.q_params_intra,
+        ff_params.q_params_solv,
+        ff_params.lj_params,
+        ff_params.lj_params_intra,
+        ff_params.lj_params_solv,
+        lamb,
+    )
+
+    # for host-guest topology, need to modify parameters for ligand-water and ligand-ligand nonbonded potentials
+    # separately
+    assert isinstance(nb_pot, SummedPotential)
+    ww_pot, lw_pot, ll_pot = nb_pot.potentials
+    assert isinstance(ww_pot, Nonbonded)
+    assert isinstance(lw_pot, NonbondedInteractionGroup)
+    assert isinstance(ll_pot, NonbondedPairListPrecomputed)
+
+    ww_params, lw_params, ll_params = nb_pot.unflatten_params(nb_params)
+    lw_params = cast(jax.Array, lw_params)
+    ll_params = cast(jax.Array, ll_params)
+
+    lw_params = lw_params.at[decouple_atom_idxs, 3].set(lamb)
+
+    decouple_atom = np.isin(ll_pot.idxs, decouple_atom_idxs)
+    decouple_pair = np.logical_xor(decouple_atom[:, 0], decouple_atom[:, 1])
```

Comment:
Intent clarification: I think this encodes that _all pairs_ of atoms in the set `decouple_atom_idxs` are decoupled from each other at the same rate that the set `decouple_atom_idxs` are decoupled (as a unit) from the rest of the environment. Is that correct?

(If there's uncertainty here, might be useful to loosen the signature to something more explicit and flexible like `intramol_atom_pairs_to_decouple: List[Pair] = ..., atoms_to_decouple_from_env: List[Int] = ...` or similar, and possibly leave a TODO comment to revisit more flexible scheduling in the future. (I guess, this family of thermodynamic states has `n * (n-1) / 2 + n` free parameters, if we control each intramol pair's w coord independently, and each atom's env-interaction w coord independently. It is not a requirement that HREX needs to look at a 1D sequence of such states.))

(If there's further uncertainty about what other "softening" may be useful to apply to this example (e.g. optionally reducing a PeriodicTorsion force constant) leaving a note in docstring or similar will be a helpful prompt when revisiting.)

Reply:
> I think this encodes that all pairs of atoms in the set decouple_atom_idxs are decoupled from each other at the same rate that the set decouple_atom_idxs are decoupled (as a unit) from the rest of the environment. Is that correct?

Correct.

> If there's uncertainty here, might be useful to loosen the signature to something more explicit and flexible like intramol_atom_pairs_to_decouple: List[Pair] = ..., atoms_to_decouple_from_env: List[Int] = ... or similar, and

This seems cleaner and easier to understand. Will investigate.

> possibly leave a TODO comment to revisit more flexible scheduling in the future.

Will do.

Reply:
Refined the transformation protocol to decouple only specified pairs (`[16, 17]` and `[12, 21]`) in https://github.com/proteneer/timemachine/pull/1170/commits/c86c5344d2961c12c5b20b7d8dc8f5790a1eab19

Added note and todo re revisiting more flexible decoupling schedule in https://github.com/proteneer/timemachine/pull/1170/commits/1a438cce69c583b31e47364159ae70d14463539d